### ISSUE TITLE
Server Error 500 when clicking unsubscribe link

### ISSUE THREAD

**Issue by MauriceW2** [mautic/mautic#7255]
## Bug Description
When clicking the unsubscribe link in an email, a server error 500 is returned.

| Q   | A
| --- | ---
| Mautic version | 2.15.0
| PHP version | 7.1
| Browser | Chrome 71.0.3578.98 (Official build) (64-bits)

### Steps to reproduce
1. 
2. 
 
### Log errors
[2019-02-17 00:35:56] mautic.CRITICAL: Uncaught PHP Exception Symfony\Component\Debug\Exception\FatalThrowableError: "Call to a member function getId() on null" at /var/www/clients/client2/web3/web/mautic/app/bundles/EmailBundle/Controller/PublicController.php line 183 {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableError(code: 0): Call to a member function getId() on null at /var/www/clients/client2/web3/web/mautic/app/bundles/EmailBundle/Controller/PublicController.php:183)"} []

**Comment by npracht** [comment]
Hi @MauriceW2 can you please share the HTML of your email in mautic and the email in your mailbox ?
Just the part where your unsubscribe link is available.

**Comment by MauriceW2** [comment]
Hi @npracht of course.

**This is the HTML in Mautic:**
` Click <a href="{unsubscribe_url}" rel="noopener noreferrer" style="font-size: 11px;" target="_blank"><u>here</u></a> to unsubscribe`

**This is the HTML result in the email itself:**
`Click <a style=3D"font-size: 11px;" href=3D"=
https://mautic.trancewerk.nl/email/unsubscribe/5c69659cbab8a565690945" rel=
=3D"noopener noreferrer" target=3D"_blank"><u>here</u></a> to unsubscribe=.`


I have no idea where the 3Ds are coming from or what they mean.
When I click the link, the web browser opens:
`https://mautic.trancewerk.nl/email/unsubscribe/5c69659cbab8a565690945`

So that seems correct.

However, It generates the "Uh oh!..." Error500.

**Comment by MauriceW2** [comment]
I have some additional info. I just created a new contact and in the first email sent, click the unsubscribe link. Now I see that a PART of the page where I'm redirected generates a server error 500. Perhaps it's related to the preference center.

![screenshot from 2019-02-18 12-11-14](https://user-images.githubusercontent.com/6311149/52947350-b5c11c00-3376-11e9-9b20-33382ab5c47d.png)

**Comment by MauriceW2** [comment]
And when I delete the contact, and **_then_** click the unsubscribe link, I get a 'full' Server error 500.

**Comment by npracht** [comment]
I think you preference center option are incorrect. Check you landing page that is used as preference center and in the global configuration of your instance, please be sure to enable the rights slots of the preference center (you may need to adapt your landing page depending on your modification).

**Comment by colourfulowl** [comment]
This is a none bug and is being investigated, no? It's to do with you being signed into Mautic. Whilst signed in it doesn't track you / gives you no ID - producing the above error.

**Comment by MauriceW2** [comment]
@npracht, I've checked my pref. center options and as far as I can see, they all set correctly.

- I've created a Mautic Landing Page and enabled 'Is Preference Center'
- In the builder I've added the slot types 'SEGMENTLIST' and 'SAVEPREFSBUTTON'.
- The Landing Page is published

- Then in the e-mail(s), I've selected this Preference center page.
- In the builder I've added the unsubscribe email token {unsubscribe_url}

Do I need to do more/other?

@colourfulowl I'm NOT signed into Mautic when I click the unsubscribe link. I'm receiving email and testing on a different computer. So this bug is not related to me being logged in.

For testing I've copied the complete unsubscribe URL to the computer where I'm logged into Mautic as well, and there it generates the exact same error. Even in an incognito window.

**Comment by colourfulowl** [comment]
Your preference centre landing page, did you code it or are you using the builder? 

If you coded it can you provide a copy of the code? If it's done using the builder, can you provide us with a preview link?

It's a bit weird that your getting a 500 error when you can see some of the page. We need to see what element of the page is throwing out that 500 server error.

**Comment by MauriceW2** [comment]
@colourfulowl I just used the builder.

This is the preview link: https://mautic.trancewerk.nl/mautic/page/preview/1

It's weird, indeed. The heaader 'Message Preferences'is visible and so is the contact name + a string 'Please use the form below to set your message preferences.'.

Below that, the actual form should appear. And that's the part that shows the error. (I think)

**Comment by MauriceW2** [comment]
When I click the unsubscribe link, these two line appear in the Mautic log:

`[2019-02-19 14:23:18] mautic.WARNING: PHP Warning - Invalid argument supplied for foreach() - in file /var/www/clients/client2/web3/web/mautic/app/bundles/EmailBundle/Views/Lead/preference_options.html.php - at line 48 [] []`

`[2019-02-19 14:23:18] mautic.CRITICAL: Uncaught PHP Exception Symfony\Component\Debug\Exception\FatalThrowableError: "Type error: Argument 1 passed to Symfony\Bundle\FrameworkBundle\Templating\Helper\FormHelper::row() must be an instance of Symfony\Component\Form\FormView, null given, called in /var/www/clients/client2/web3/web/mautic/app/bundles/EmailBundle/Views/Lead/preference_options.html.php on line 100" at /var/www/clients/client2/web3/web/mautic/vendor/symfony/framework-bundle/Templating/Helper/FormHelper.php line 169 {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableError(code: 0): Type error: Argument 1 passed to Symfony\\Bundle\\FrameworkBundle\\Templating\\Helper\\FormHelper::row() must be an instance of Symfony\\Component\\Form\\FormView, null given, called in /var/www/clients/client2/web3/web/mautic/app/bundles/EmailBundle/Views/Lead/preference_options.html.php on line 100 at /var/www/clients/client2/web3/web/mautic/vendor/symfony/framework-bundle/Templating/Helper/FormHelper.php:169)"} []`

**Comment by colourfulowl** [comment]
Umm. It looks like this might be a configuration issue as the page it's loading isn't the preference center page you want it to be. Can you please make sure all of the below is completed. I've looked at the code on both pages and can't see anything out of the ordinary.

Can you go to Configuration > Email Settings and scroll down to Unsubscribe Settings. 

The first option (Show contact preference settings) needs to be checked as 'Yes'.  Right now your preference centre isn't being loaded at all, instead it loading the default.

![mautic](https://user-images.githubusercontent.com/41902603/53019481-baf09a80-344c-11e9-88f7-fe4f7ba1f5f0.PNG)

If this is done, you need to check that the preference page is selected within the email itself. 

Head over to the email you're sending, and make sure that 'Preference center page' is directed to the page you want it to be.

![mautic2](https://user-images.githubusercontent.com/41902603/53019632-16228d00-344d-11e9-8482-4df1b7d6d1d5.PNG)

If you don't see any listed, make sure you have 'Is preference center' marked as 'Yes' on the landing page. This should add a cog next to your pages name. 

![mautic3](https://user-images.githubusercontent.com/41902603/53019743-5550de00-344d-11e9-90f6-5674528a9a57.PNG)
![mautic4](https://user-images.githubusercontent.com/41902603/53019747-56820b00-344d-11e9-8f59-04754f3c6513.PNG)

**Comment by MauriceW2** [comment]
I took all the steps and this seems to work. Great, we're getting somewhere now :-)

This is the page when I click the unsubscribe link now: 
https://mautic.trancewerk.nl/email/unsubscribe/5c6c0f687a4f1412208183

However, I'm not seeing anything I setup in the Landing Page on the actual unsubscribe page, after clicking the link. It looks like a default page.

The only difference I had was that **I had set all 6 settings** in the unsubscribe settings to 'Yes'.

I've checked some further and found that when I set 'Show contact's preferred channel option' the server error 500 appears.

When I leave that setting to 'No', the error does not appear.
